packages/abstraxion-core/src/AbstraxionAuth.ts

```diff
@@ -248,98 +296,105 @@ export class AbstraxionAuth {
-            currentUrl.searchParams.delete("granter");
-            history.pushState({}, "", currentUrl.href);
-            return true;
-          }
-        } else {
-          // No grants found, retry with exponential backoff
-          if (retries < maxRetries) {
-            const delay = Math.pow(2, retries) * 1000;
-            setTimeout(poll, delay);
-            retries++;
-          } else {
-            console.error("Max retries exceeded, giving up.");
-            return false;
-          }
+          return true;
         }
+        // Retry if no grants found
+        throw new Error("No grants found.");
       } catch (error) {
-        console.error("Error while polling:", error);
-        // Retry immediately in case of network error
-        if (retries < maxRetries) {
-          setTimeout(poll, 1000);
-          retries++;
-        } else {
-          console.error("Max retries exceeded, giving up.");
-          throw error;
-        }
+        console.warn("Error fetching grants: ", error);
+        const delay = Math.pow(2, retries) * 1000;
+        await new Promise((resolve) => setTimeout(resolve, delay));
+        retries++;
       }
-    };
-
-    await poll();
+    }
+    console.error("Max retries exceeded, giving up.");
+    return false;
   }
 
-  logout() {
+  /**
+   * Wipe persisted state and instance variables.
+   */
+  logout(): void {
     localStorage.removeItem("xion-authz-temp-account");
     localStorage.removeItem("xion-authz-granter-account");
-    this.signArbWallet = undefined;
     this.abstractAccount = undefined;
     this.triggerAuthStateChange(false);
   }
 
-  // Do we want to build this out more?
-  async authenticate() {
+  /**
+   * Authenticates the user based on the presence of a local keypair and a granter address.
+   * If a local keypair and granter address are found, sets the abstract account and triggers authentication state change.
+   * This method is typically called to authenticate the user and persist state between renders.
+   */
+  async authenticate(): Promise<void> {
     const keypair = await this.getLocalKeypair();
     const granter = this.getGranter();
 
     if (keypair && granter) {
-      this.abstractAccount = this.signArbWallet;
+      this.abstractAccount = keypair;
       this.triggerAuthStateChange(true);
     }
   }
 
-  // Handle different possible cases of logging in
-  async login() {
+  /**
+   * Initiates the login process for the user.
+   * Checks if a local keypair and granter address exist, either from URL parameters or localStorage.
+   * If both exist, polls for grants and updates the authentication state if successful.
+   * If not, generates a new keypair and redirects to the dashboard for grant issuance.
+   *
+   * @returns {Promise<void>} - A Promise that resolves once the login process is complete.
+   * @throws {Error} - If the login process encounters an error.
+   */
+  async login(): Promise<void> {
```

Comment:
This method needs to be resilient to being called several times while the login process is in progress.

Reply:
Definitely agree. Maybe some instance state that a login has already been triggered. Any ideas on actions? Just simply return? Maybe a warning that a login is already in progress, etc?